binderhub/repoproviders.py

```diff
@@ -83,6 +91,138 @@ def get_build_slug(self):
         return '{user}-{repo}'.format(user='Rick', repo='Morty')
 
 
+class GitRepoProvider(RepoProvider):
+    """Bare bones git repo provider.
+
+    Users must provide a spec of the following form.
+
+    <url-escaped-namespace>/<resolved_ref>
+
+    eg:
+    https%3A%2F%2Fgithub.com%2Fjupyterhub%2Fzero-to-jupyterhub-k8s/f7f3ff6d1bf708bdc12e5f10e18b2a90a4795603
+    """
+
+    name = Unicode("Git")
+
+    def __init__(self, *args, **kwargs):
+        super().__init__(*args, **kwargs)
+        url, resolved_ref = self.spec.rsplit('/', 1)
+        self.repo = urllib.parse.unquote(url)
+        if not resolved_ref:
+            raise ValueError("`resolved_ref` must be specified as a query parameter for the basic git provider")
+        self.sha1_validate(resolved_ref)
+        self.resolved_ref = resolved_ref
+
+    @gen.coroutine
+    def get_resolved_ref(self):
+        return self.resolved_ref
+
+    def get_repo_url(self):
+        return self.repo
+
+    def get_build_slug(self):
+        parsed = urllib.parse.urlparse(self.repo)
+        parts = [parsed.netloc]
+        parts.extend(parsed.path.strip('/').split('/'))
+        # escape the name and replace dashes with something else.
+        return '-'.join(p.replace('-', '_-') for p in parts)
+
+
+class GitLabRepoProvider(RepoProvider):
+    """GitLab provider.
+
+    GitLab allows nested namespaces (eg. root/project/component/repo) thus we need to urlescape the namespace of this
+    repo.  Users must provide a spec that matches the following form.
+
+    <url-escaped-namespace>/<unresolved_ref>
+
+    eg:
+    group%2Fproject%2Frepo/master
+    """
+
+    name = Unicode('GitLab')
+    hostname = Unicode('gitlab.com')
+
+    access_token = Unicode(config=True,
+        help="""Gitlab OAuth2 access token for authentication with the Gitlab API
+
+        For use with client_secret.
+        Loaded from GITLAB_ACCESS_TOKEN env by default.
+        """
+    )
+    @default('access_token')
+    def _access_token_default(self):
+        return os.getenv('GITLABV_ACCESS_TOKEN', '')
+
+    private_token = Unicode(config=True,
+        help="""Gitlab private token for authentication with the Gitlab API
+
+        Loaded from GITLAB_PRIVATE_TOKEN env by default.
+        """
+    )
+    @default('private_token')
+    def _private_token_default(self):
+        return os.getenv('GITLAB_PRIVATE_TOKEN', '')
```

Comment:
@minrk could something similar to this be used for private github repos? On the assumption that the person deploying the binderhub would have those tokens for a specific private repo already on the machine from which they deploy?

Reply:
Yup, something similar to this needs to happen to support private repos on github :)